resources/top_level_overview.markdown

```diff
@@ -0,0 +1,283 @@
+zero or one output values are produced by any _Reduced_ invocation.
+(The list of values are passed by an iterator to deal with memory availability.)
+
+The above computational process has the following type associations:
+
+```
+map:    (k1, v1)       -> list(k2, v2)
+
+reduce: (k2, list(v2)) -> list(v2)
+```
+
+That is, a map task, which is a copy of a user-defined function, accepts
+key-value pairs, `(k1, v1)`, and outputs `(k2, v2)`, representative of the
+intermediate value pairs, `I`.  These intermediate pairs are not necessarily of
+the same type as the input pairs.  The reducer then accepts all keys, `k2`, and
+the corresponding list of values, `list(v2)`.  The reducer then performs some
+user-defined operation on the list, and produces some output.
+
+
+#### Steps in a MapReduce Process:
+
+The following is a simplification of the steps outlined in
+_Map Reduce: Simplified Data Processing on Large Clusters_.  After we break down
+the process, we'll explore how Hadoop handles each of the following:
+
+1. The MapReduce library splits the input into _M_ chunks, and starts up copies
+of the program on multiple machines.
+
+2. The _master_ fork assigns tasks to the _worker_ nodes who are just running
+copies of the _user program_.  There are `M` map tasks, and `R` reduce tasks
+that the _master_ must assign. Note that a worker node takes on the role of
+either map, or reduce. Typically, there are many more map nodes than reduce
+nodes.
+
+3. A _worker_ assigned a map task, reads in the contents of the corresponding
+input split.  The _worker_ parses (key, values) pairs and passes it to the map
+function. The _intermediate_ (key, value) pairs produced by the mapping are
+buffered into memory.
+
+4. The local disk it partitioned into `R` regions by the _partitioning
+function_. The buffered _intermediate_ pairs produced by a map worker are
+periodically written to these partitions, with the particular location of these
+writes sent back to `master` for use by the reduce workers.
+
+5. When the `master` notifies a _reduce_ worker about a particular location of
+_intermediate_ key-value pairs, it reads the buffered data from the local
+disks of the map workers. When a reduce worker has read all intermediate data,
+it sorts it by key so that all of the same keys are grouped together. The
+sorting is needed because many different keys can map to the same reduce task.
+If the amount of intermediate data is too large to fit in memory, an external
+sort is used.
+
+6. The reduce worker iterates over the sorted data and for each unique
+intermediate key encountered, it passes the key and the corresponding set of
+intermediate values to the user’s _reduce_ function. The output of the reduce
+function is appended to a final output file for this reduce partition.
+
+7. When all map tasks and reduce tasks have been completed, the call to
+MapReduce reaches a return statement back to user code.
```

Comment:
would be nice to say here what, at that point, is on disk. A single output file? An output file for each Reducer?